lib/Core/BuildEngine.cpp

```diff
@@ -931,31 +942,97 @@ class BuildEngineImpl {
-    if (!taskInfos.empty()) {
-      reportCycle(buildKey);
-      return false;
+        // We didn't do any work, and we have nothing more we can/need to do.
+        break;
+      }
     }
 
     return true;
   }
 
-  /// Report the cycle which has called the engine to be unable to make forward
+  /// Attempt to resolve a cycle which has called the engine to be unable to make forward
   /// progress.
   ///
   /// \param buildKey The key which was requested to build (the reported cycle
   /// with start with this node).
-  void reportCycle(const KeyType& buildKey) {
+  /// \returns True if the engine should try to proceed, false if the build the could not
+  /// be broken.
+  bool resolveCycle(const KeyType& buildKey) {
     // Take all available locks, to ensure we dump a consistent state.
     std::lock_guard<std::mutex> guard1(taskInfosMutex);
     std::lock_guard<std::mutex> guard2(finishedTaskInfosMutex);
 
+    std::vector<Rule*> cycleList = findCycle(buildKey);
+    assert(!cycleList.empty());
+
+    if (delegate.shouldResolveCycle(cycleList)) {
+      TracingInterval i(EngineQueueItemKind::BreakingCycle);
+
+      for (auto ruleIt = cycleList.rbegin(); ruleIt != cycleList.rend(); ruleIt++) {
+        auto& ruleInfo = getRuleInfoForKey((*ruleIt)->key);
+
+        // If this rule is scanning, try to force a rebuild to break the cycle
+        if (ruleInfo.isScanning()) {
+          if (trace) {
+            trace->cycleForceRuleNeedsToRun(&ruleInfo.rule);
+          }
+          for (auto it = ruleInfosToScan.begin(); it != ruleInfosToScan.end(); it++) {
+            if (it->ruleInfo->keyID == ruleInfo.keyID) {
+              ruleInfosToScan.erase(it);
+            }
+          }
+          finishScanRequest(ruleInfo, RuleInfo::StateKind::NeedsToRun);
+          return true;
+        }
+
+        // Check if this rule has a (potentially) valid previous result and if so
+        // try to provide it to the node requesting it to break the cycle.
+        if (ruleInfo.isInProgressWaiting() && ruleInfo.result.builtAt != 0) {
+          auto* taskInfo = ruleInfo.getPendingTaskInfo();
+
+          // find downstream node in the cycle that wants this input
+          auto& nextRuleInfo = getRuleInfoForKey((*std::next(ruleIt))->key);
+
+          // find the corresponding request on this task
+          for (auto it = taskInfo->requestedBy.begin(); it != taskInfo->requestedBy.end(); it++) {
+            if (it->taskInfo->forRuleInfo->keyID == nextRuleInfo.keyID) {
+              // supply the prior value to the node
+              if (trace) {
+                trace->cycleSupplyPriorValue(&ruleInfo.rule, it->taskInfo->task.get());
+              }
+              it->forcePriorValue = true;
+              finishedInputRequests.insert(finishedInputRequests.end(), *it);
+
+              // remove this request from the task info
+              taskInfo->requestedBy.erase(it);
+              return true;
+            }
+          }
```

Comment:
It would be a serious error if this loop fell through, right?

I would suggest phrasing this as a search for the matching node (maybe via a helper function), which must succeed, then an assert, then the logic to break that node.